api/src/main/java/org/projectnessie/tools/catalog/migration/api/CatalogMigrator.java

```diff
@@ -0,0 +1,200 @@
+  public abstract Catalog targetCatalog();
+
+  /** Delete the table entries from the source catalog after successful registration. */
+  public abstract boolean deleteEntriesFromSourceCatalog();
+
+  /** Enable the stacktrace in logs in case of failures. */
+  @Value.Default
+  public boolean enableStacktrace() {
+    return false;
+  }
+
+  private static final Logger LOG = LoggerFactory.getLogger(CatalogMigrator.class);
+  private final ImmutableCatalogMigrationResult.Builder resultBuilder =
+      ImmutableCatalogMigrationResult.builder();
+  private final Set<Namespace> processedNamespaces = new HashSet<>();
+
+  /**
+   * Get the table identifiers which matches the regular expression pattern input from all the
+   * namespaces.
+   *
+   * @param identifierRegex regular expression pattern. If null, fetches all the table identifiers
+   *     from all the namespaces.
+   * @return List of table identifiers.
+   */
+  public List<TableIdentifier> getMatchingTableIdentifiers(String identifierRegex) {
+    Catalog sourceCatalog = sourceCatalog();
+    if (!(sourceCatalog instanceof SupportsNamespaces)) {
+      throw new UnsupportedOperationException(
+          String.format(
+              "source catalog %s doesn't implement SupportsNamespaces to list all namespaces.",
+              sourceCatalog.name()));
+    }
+    LOG.info("Collecting all the namespaces from source catalog...");
+    List<Namespace> namespaces = ((SupportsNamespaces) sourceCatalog).listNamespaces();
+    Predicate<TableIdentifier> matchedIdentifiersPredicate;
+    if (identifierRegex == null) {
+      LOG.info("Collecting all the tables from all the namespaces of source catalog...");
+      matchedIdentifiersPredicate = tableIdentifier -> true;
+    } else {
+      LOG.info(
+          "Collecting all the tables from all the namespaces of source catalog"
+              + " which matches the regex pattern:{}",
+          identifierRegex);
+      Pattern pattern = Pattern.compile(identifierRegex);
+      matchedIdentifiersPredicate =
+          tableIdentifier -> pattern.matcher(tableIdentifier.toString()).matches();
+    }
+    return namespaces.stream()
+        .filter(Objects::nonNull)
+        .flatMap(
+            namespace ->
+                sourceCatalog.listTables(namespace).stream().filter(matchedIdentifiersPredicate))
+        .collect(Collectors.toList());
+  }
+
+  /**
+   * Register or Migrate tables from one catalog(source catalog) to another catalog(target catalog).
+   *
+   * <p>Users must make sure that no in-progress commits on the tables of source catalog during
+   * registration.
+   *
+   * @param identifiers List of table identifiers to register or migrate
+   * @return {@code this} for use in a chained invocation
+   */
+  public CatalogMigrator registerTables(List<TableIdentifier> identifiers) {
+    Preconditions.checkArgument(identifiers != null, "Identifiers list is null");
```

Comment:
CLI level handling is already there. When called from CLI, this will never be empty or null. 
But the `CatalogMigrator` can be used as API also. Without a CLI. That API needs input validation and these are for that.